### PR TITLE
Add documentation comments for several types

### PR DESCRIPTION
**Pull request by emesterhazy**
These comments are intended to make it easier for new developers to get up to speed with the project. This is just a starting point... there are other types and functions that could benefit from documentation.

### CODE REVIEW

File: lib/src/backend.rs [hunk 1]
```diff
@@ -331,6 +333,7 @@ pub fn make_root_commit(root_change_id: ChangeId, empty_tree_id: TreeId) -> Comm
     }
 }
 
+/// Defines the interface for backends that can be used to store a repository.
```
Comment (martinvonz) [inline lib/src/backend.rs:336]: nit: Since we have several kinds of backends, say "for commit backends" here

File: lib/src/object_id.rs [hunk 2]
```diff
@@ -103,14 +103,19 @@ macro_rules! impl_id_type {
 
 pub(crate) use {id_type, impl_id_type};
 
+/// Stores bytes representing the minimum length unique hexadecimal prefix of an
+/// identifier.
```
Comment (martinvonz) [inline lib/src/object_id.rs:107]: I'm not sure we actually use it for converting a full id to a prefix. I know we use it for the reverse. So the type is used more generally to represent a hex prefix whether it has the minimum length or not.

File: lib/src/object_id.rs [hunk 3]
```diff
@@ -103,14 +103,19 @@ macro_rules! impl_id_type {
...
 pub struct HexPrefix {
-    // For odd-length prefix, lower 4 bits of the last byte is padded with 0
+    // For odd-length prefixes, the lower 4 bits of the last byte are
+    // zero-filled (e.g. the prefix "abc" is stored in two bytes as 0xabc0).
     min_prefix_bytes: Vec<u8>,
     has_odd_byte: bool,
 }
 
 impl HexPrefix {
+    /// Decodes `prefix` to bytes and returns a new `HexPrefix`. Returns `None`
```
Comment (martinvonz) [inline lib/src/object_id.rs:117]: nit: I feel like the "Decodes `prefix` to bytes" bit is an implementation detail that's not important to the caller.

File: lib/src/object_id.rs [hunk 4]
```diff
@@ -224,6 +231,12 @@ mod tests {
 
         let prefix = HexPrefix::new("123").unwrap();
         assert_eq!(prefix.min_prefix_bytes(), b"\x12\x30");
+
+        let bad_prefix = HexPrefix::new("0x123");
+        assert_eq!(bad_prefix, None);
+
+        let bad_prefix = HexPrefix::new("foobar");
+        assert_eq!(bad_prefix, None);
```
Comment (martinvonz) [inline lib/src/object_id.rs:239]: nit: should be a separate commit

**Review by martinvonz — APPROVED** [review]
Thanks!

I don't think it applies to any of these files, but if you manage to document all public symbols in a file, please remove `#![allow(missing_docs)]` to prevent regressions.

File: lib/src/backend.rs (hunk 1 above)
Comment (emesterhazy) [reply lib/src/backend.rs:336]: Good point, thanks!

File: lib/src/object_id.rs (hunk 2 above)
Comment (emesterhazy) [reply lib/src/object_id.rs:107]: I reworded this completely. Let me know if it's better.

File: lib/src/object_id.rs (hunk 3 above)
Comment (emesterhazy) [reply lib/src/object_id.rs:117]: I reworded this slightly. I think the caller just needs to know that `None` is returned if "prefix" can't be decoded from hex to bytes.

File: lib/src/object_id.rs (hunk 4 above)
Comment (emesterhazy) [reply lib/src/object_id.rs:239]: Will do!

**Review by martinvonz — APPROVED** [review]

File: lib/src/object_id.rs [hunk 5]
```diff
@@ -103,14 +103,19 @@ macro_rules! impl_id_type {
 
 pub(crate) use {id_type, impl_id_type};
 
+/// An identifier prefix (typically from a type implementing the [`ObjectId`]
+/// trait) with facilities for converting between bytes and a hex string.
 #[derive(Debug, Clone, PartialEq, Eq)]
 pub struct HexPrefix {
-    // For odd-length prefix, lower 4 bits of the last byte is padded with 0
+    // For odd-length prefixes, the lower 4 bits of the last byte are
+    // zero-filled (e.g. the prefix "abc" is stored in two bytes as 0xabc0).
```
Comment (yuja) [inline lib/src/object_id.rs:111]: nit: `b"\xab\xc0" or simply `"abc0"` to avoid endianness stuff?

**Review by yuja — APPROVED** [review]

File: lib/src/object_id.rs (hunk 5 above)
Comment (emesterhazy) [reply lib/src/object_id.rs:111]: Thanks, good point. I'll change it to "abc".

Comment (emesterhazy) [reply lib/src/object_id.rs:111]: Sorry, typo. Changing to "abc0".
